src/numerics/include/metaphysicl/dualnumber_surrogate.h

```diff
@@ -0,0 +1,316 @@
+    derivatives[di] = &dn.derivatives()[di](std::forward<Args>(args)...);
+}
+
+template <typename T, std::size_t N>
+template <typename T2, class... Args>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumber<T2, NumberArray<N, T2>> && dn,
+                                                      Args &&... args)
+  : value(dn.value()(std::forward<Args>(args)...))
+{
+  for (decltype(N) di = 0; di < N; ++di)
+    derivatives[di] = &dn.derivatives()[di](std::forward<Args>(args)...);
+}
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumberSurrogate<T, N> & dns)
+  : value(dns.value), derivatives(dns.derivatives)
+{
+}
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(const DualNumberSurrogate<T, N> & dns)
+  : value(dns.value), derivatives(dns.derivatives)
+{
+}
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumberSurrogate<T, N> && dns)
+  : value(dns.value), derivatives(dns.derivatives)
+{
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator=(DualNumberSurrogate<T2, N> & dns)
+{
+  value = dns.value;
+  for (decltype(N) i = 0; i < N; ++i)
+  {
+    if (dns.derivatives[i])
+      *derivatives[i] = *dns.derivatives[i];
+    else
+      *derivatives[i] = 0;
+  }
+  return *this;
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator=(const DualNumberSurrogate<T2, N> & dns)
+{
+  value = dns.value;
+  for (decltype(N) i = 0; i < N; ++i)
+  {
+    if (dns.derivatives[i])
+      *derivatives[i] = *dns.derivatives[i];
+    else
+      *derivatives[i] = 0;
+  }
+  return *this;
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator=(DualNumberSurrogate<T2, N> && dns)
+{
+  value = dns.value;
+  for (decltype(N) i = 0; i < N; ++i)
+  {
+    if (dns.derivatives[i])
+      *derivatives[i] = *dns.derivatives[i];
+    else
+      *derivatives[i] = 0;
+  }
+  return *this;
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator=(const T2 & in_value)
+{
+  value = in_value;
+  for (decltype(N) i = 0; i < N; ++i)
+    *derivatives[i] = 0;
+  return *this;
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator=(const DualNumber<T2, NumberArray<N, T2>> & in_dn)
+{
+  value = in_dn.value();
+  for (decltype(N) i = 0; i < N; ++i)
+    *derivatives[i] = in_dn.derivatives()[i];
+  return *this;
+}
+
+template <typename T, std::size_t N>
+template <typename T2>
+inline DualNumberSurrogate<T, N> &
+DualNumberSurrogate<T, N>::operator+=(const DualNumberSurrogate<T2, N> & dns)
```

Comment:
These operators, on the other hand, are where I would have just given up and used stupid macro tricks.  Your way is probably better.